### ISSUE TITLE
What should igraph_ring() return for n=1 and 2?

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1870]
When asking close "closed rings", `igraph_ring()` returns:

 - for `n=1`, the singleton graph with no edges
 - for `n=2`, two vertices connected by a single edge

<img width="389" alt="image" src="https://user-images.githubusercontent.com/1212871/144755320-dee3b70d-1f45-428d-a92a-60c8a67d07a2.png">

Should it instead return:

 - for `n=1`, a singleton graph with a self-loop
 - for `n=2`, two vertices connected by two edges

<img width="389" alt="image" src="https://user-images.githubusercontent.com/1212871/144755326-4c554fcd-6fe3-4a28-ba6a-e1d0242c8174.png">

This is what I would expect for the "1-cycle" and the "2-cycle".

The reason why it behaves this way is that it's just a wrapper for `igraph_lattice()`. I would rewrite the function fro scratch and make sure that modifications to `igraph_lattice()` would not affect it.

**Comment by ntamas** [comment]
Personally, I'd prefer the second figure presented above instead of the first. Any objections?

If we do this, this is a breaking change so it should go into 0.10.

**Comment by szhorvat** [comment]
It seems we have a decision, `n=1` and `n=2` cases should be modified as in the second figure. This is for 0.10 only. Docs need updating too, including in all high-level interfaces where it is claimed that this is a wrapper for `igraph_lattice()`.

**Comment by szhorvat** [comment]
Tagging as `good first issue` because this is simple to fix, and can be an educational reimplementation of this functionality.

**Comment by ayushagr2002** [comment]
Hi. I would like to work on this issue. But i am not able to install igraph using the instructions in the manual. It shows the following:
` CMake Error at src/CMakeLists.txt:20 (bison_target):
   Unknown CMake command "bison_target".`
Could you help me with this?

**Comment by ntamas** [comment]
You need to install `flex` and `bison` - these are generators that generate parsers for some of the file formats that igraph supports. Without them, CMake won't provide a `bison_target` command so it won't be able to generate the build system for you.

**Comment by ayushagr2002** [comment]
Thank you for the solution. Now for the last step of installation, when i am trying to execute `cmake --install .`, i get the following error:
```
-- Install configuration: "Release"
-- Installing: /usr/local/lib/libigraph.a
CMake Error at src/cmake_install.cmake:56 (file):
  file INSTALL cannot copy file
  "/home/ayush/igraphres/igraph/build/src/libigraph.a" to
  "/usr/local/lib/libigraph.a": Permission denied.
Call Stack (most recent call first):
  cmake_install.cmake:47 (include)
```
Can you help me with this please?

**Comment by szhorvat** [comment]
I suggest that you do not install the library at the default, system-wide location, which would require root access. Instead, install it in a directory of your choosing, as described here:

https://igraph.org/c/html/latest/igraph-Installation.html#igraph-Installation-general-build-instructions

Set `CMAKE_INSTALL_PREFIX` to the install location.

When using an already installed igraph, [as described here](https://igraph.org/c/html/latest/igraph-Tutorial.html#tut-lesson-1-compiling-with-cmake), set `CMAKE_PREFIX_PATH` to the installation location.

**Comment by szhorvat** [comment]
> Edit: I solved this issue using sudo.

As I said in my message above, that is not at all a good idea. Do not install a development version at a system-wide location.

**Comment by szhorvat** [comment]
Regarding the requirement for `bison` and `flex`, it is documented here:

https://igraph.org/c/html/latest/igraph-Installation.html#igraph-Installation-prerequisites

Feedback about the installation instructions is welcome. If something is missing, let us know and we will try to add it.

**Comment by ayushagr2002** [comment]
> > Edit: I solved this issue using sudo.
> 
> As I said in my message above, that is not at all a good idea. Do not install a development version at a system-wide location.

Got it. Thanks.

**Comment by ayushagr2002** [comment]
> Regarding the requirement for `bison` and `flex`, it is documented here:
> 
> https://igraph.org/c/html/latest/igraph-Installation.html#igraph-Installation-prerequisites
> 
> Feedback about the installation instructions is welcome. If something is missing, let us know and we will try to add it.

I believe the installation instructions can be made more detailed. We could make a guide for the user to check if all the necessary requirements are met and if not how to install them correctly. I can make an improved version of installation guide for better convenience to users, if you approve.

**Comment by szhorvat** [comment]
> I believe the installation instructions can be made more detailed. We could make a guide for the user to check if all the necessary requirements are met and if not how to install them correctly. I can make an improved version of installation guide for better convenience to users, if you approve.

_Feedback_ about the installation instructions is very welcome, and we will use that for future improvements. But at this stage, I think it's best if we write this critical piece of the documentation ourselves. It must cater to a very broad audience, which is difficult to achieve. Adding more details is easy, but in my experience it makes the text too long, and people will not read it in the end. Notice that the issue of both `bison` and `flex`, as well as setting an installation location, are both mentioned in the current documentation, but you seem to have missed them. Did you read the text in full? If not, perhaps it is already too long. The challenge is being concise and informative at the same time.

**Comment by ayushagr2002** [comment]
Regarding the original issue, do you expect changes in `igraph_lattice()` function only?(to handle the case for n=1 and n=2)

**Comment by szhorvat** [comment]
Let us leave `igraph_lattice()` as it is, as its behaviour is reasonable.

Instead, let us rewrite `igraph_ring()` from scratch, and make it independent of `igraph_lattice()`

**Comment by szhorvat** [comment]
I recommend you start with a skeleton draft PR, and work in public, so we can give early feedback. See the very first commit in PR #1859 for an example.

**Comment by ayushagr2002** [comment]
For the case of n=2, it is applicable for directed graphs only right? 
For the case of n=1, for undirected graphs, should the edge set be {(0,0)} or {0}?

**Comment by ayushagr2002** [comment]
[https://github.com/ayushagr2002/igraph/commit/12bba773eb2a034b99234e047b5c138d2d1f08e8](url)
Please have a look at the igraph_ring() function

**Comment by ayushagr2002** [comment]
I added the following in igraph_ring() function:
```
if(n == 1 && circular)
    {
        igraph_vector_t edges;

        //initialise edge vector and insert edge 0-0 for self loop
        igraph_vector_init(&edges,2);
        igraph_vector_set(&edges,0,0);
        igraph_vector_set(&edges,1,0);

        //create the graph
        IGRAPH_CHECK(igraph_create(graph,&edges,0,directed));
        igraph_vector_destroy(&edges);
        //IGRAPH_FINALLY_CLEAN(1);

        return IGRAPH_SUCCESS;
    }

    if(n == 2 && circular && directed)
    {
        igraph_vector_t edges;
        //initialise edge vector and insert edge 0-0 for self loop
        igraph_vector_init(&edges,4);
        igraph_vector_set(&edges,0,0);
        igraph_vector_set(&edges,1,1);
        igraph_vector_set(&edges,2,1);
        igraph_vector_set(&edges,3,0);

        //create the graph
        IGRAPH_CHECK(igraph_create(graph,&edges,0,directed));
        igraph_vector_destroy(&edges);
        //IGRAPH_FINALLY_CLEAN(1);

        return IGRAPH_SUCCESS;
    }
```

**Comment by GroteGnoom** [comment]
Could you make a draft PR?

**Comment by ayushagr2002** [comment]
I have created a draft pr. Please look into it.

**Comment by szhorvat** [comment]
Resolved by e9295c61443c354738409edf713cf768941a12b1
